test/integration/certificates/cleanup_policy_test.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+
+	"github.com/cert-manager/cert-manager/internal/apis/certmanager"
+	apiutil "github.com/cert-manager/cert-manager/pkg/api/util"
+	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	controllerpkg "github.com/cert-manager/cert-manager/pkg/controller"
+	"github.com/cert-manager/cert-manager/pkg/controller/certificates/issuing"
+	logf "github.com/cert-manager/cert-manager/pkg/logs"
+	"github.com/cert-manager/cert-manager/pkg/metrics"
+	"github.com/cert-manager/cert-manager/test/integration/framework"
+	testcrypto "github.com/cert-manager/cert-manager/test/unit/crypto"
+	"github.com/cert-manager/cert-manager/test/unit/gen"
+)
+
+// Test_CleanupPolicyChange tests change state of issuerRef on created Secret
+// when certificate cleanupPolicy is changed.
```

Comment:
I would probably have gone with using the fake Kubernetes client (named `testpkg` in cert-manager) instead of an integration tests (ie, envtest, which relies on a live apiserver). It would go much faster with a fake Kubernetes client.

I imagine you already spent a considerable amount of time on this feature, so I'd be OK if you prefer sticking to this integration test.

Reply:
I think we need to approve idea of this PR, and, if need, i will rewrite tests.

Reply:
Hi. Now that the idea has been approved, we can proceed with taking another look at this integration test.

Reply:
Hi, i'll try to rewrite tests to fake client